qa/logger-usage/src/main/java/org/elasticsearch/qa/loggerusage/ESLoggerUsageChecker.java

```diff
@@ -0,0 +1,437 @@
+            } else {
+                return super.visitMethod(access, name, desc, signature, exceptions);
+            }
+        }
+    }
+
+    private static class MethodChecker extends MethodVisitor {
+        private final String className;
+        private final Consumer<WrongLoggerUsage> wrongUsageCallback;
+
+        public MethodChecker(String className, int access, String name, String desc, Consumer<WrongLoggerUsage> wrongUsageCallback) {
+            super(Opcodes.ASM5, new MethodNode(access, name, desc, null, null));
+            this.className = className;
+            this.wrongUsageCallback = wrongUsageCallback;
+        }
+
+        @Override
+        public void visitEnd() {
+            findBadLoggerUsages((MethodNode) mv);
+            super.visitEnd();
+        }
+
+        public void findBadLoggerUsages(MethodNode methodNode) {
+            Analyzer<BasicValue> stringPlaceHolderAnalyzer = new Analyzer<>(new PlaceHolderStringInterpreter());
+            Analyzer<BasicValue> arraySizeAnalyzer = new Analyzer<>(new ArraySizeInterpreter());
+            try {
+                stringPlaceHolderAnalyzer.analyze(className, methodNode);
+                arraySizeAnalyzer.analyze(className, methodNode);
+            } catch (AnalyzerException e) {
+                throw new RuntimeException("Internal error: failed in analysis step", e);
+            }
+            Frame<BasicValue>[] stringFrames = stringPlaceHolderAnalyzer.getFrames();
+            Frame<BasicValue>[] arraySizeFrames = arraySizeAnalyzer.getFrames();
+            AbstractInsnNode[] insns = methodNode.instructions.toArray();
+            int lineNumber = -1;
+            for (int i = 0; i < insns.length; i++) {
+                AbstractInsnNode insn = insns[i];
+                if (insn instanceof LineNumberNode) {
+                    LineNumberNode lineNumberNode = (LineNumberNode) insn;
+                    lineNumber = lineNumberNode.line;
+                }
+                if (insn.getOpcode() == Opcodes.INVOKEINTERFACE) {
+                    MethodInsnNode methodInsn = (MethodInsnNode) insn;
+                    if (Type.getObjectType(methodInsn.owner).getClassName().equals(LOGGER_CLASS) == false) {
+                        continue;
+                    }
+                    if (LOGGER_METHODS.contains(methodInsn.name) == false) {
+                        continue;
+                    }
+                    Type[] argumentTypes = Type.getArgumentTypes(methodInsn.desc);
+                    BasicValue logMessageLengthObject = getStackValue(stringFrames[i], argumentTypes.length - 1); // first argument
+                    if (logMessageLengthObject instanceof PlaceHolderStringBasicValue == false) {
+                        // we should probably require that only string constants are passed as first argument
+                        // badUsages.add(new WrongLoggerUsage(className, methodNode.name, lineNumber,
```

Comment:
Yeah. If we don't then the check just silently skips the invocation. Something like "I don't understand log invocations of this type."